sqlglot/dialects/dialect.py

```diff
@@ -1907,12 +1907,19 @@ def groupconcat_sql(
 
 def build_timetostr_or_tochar(args: t.List, dialect: Dialect) -> exp.TimeToStr | exp.ToChar:
     this = seq_get(args, 0)
+    format = seq_get(args, 1)
 
-    if this and not this.type:
-        from sqlglot.optimizer.annotate_types import annotate_types
+    if this:
+        if not this.type:
+            from sqlglot.optimizer.annotate_types import annotate_types
+
+            annotate_types(this, dialect=dialect)
 
-        annotate_types(this, dialect=dialect)
-        if this.is_type(*exp.DataType.TEMPORAL_TYPES):
+        from sqlglot.dialects import Snowflake
+
+        if this.is_type(*exp.DataType.TEMPORAL_TYPES) or (
+            isinstance(format, exp.Literal) and format.name in Snowflake.TIME_MAPPING
```

Comment:
Should we leave a comment here stating the present assumptions? Like that this helper is only used for Snowflake and Oracle which are expected to be compatible, etc etc...